cmd/collect_signals/main.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/ossf/criticality_score/cmd/collect_signals/collector"
+	"github.com/ossf/criticality_score/cmd/collect_signals/github"
+	"github.com/ossf/criticality_score/cmd/collect_signals/githubmentions"
+	"github.com/ossf/criticality_score/cmd/collect_signals/result"
+	"github.com/ossf/criticality_score/internal/githubapi"
+	"github.com/ossf/criticality_score/internal/logflag"
+	"github.com/ossf/criticality_score/internal/outfile"
+	"github.com/ossf/scorecard/v4/clients/githubrepo/roundtripper"
+	sclog "github.com/ossf/scorecard/v4/log"
+	log "github.com/sirupsen/logrus"
+	//csvutil (it's docs are heaps better!)
+)
+
+const defaultLogLevel = log.InfoLevel
+
+var logFlag = logflag.Level(defaultLogLevel)
+
+func init() {
+	flag.Var(&logFlag, "log", "set the `level` of logging.")
+	outfile.DefineFlags(flag.CommandLine, "force", "append", "OUT_FILE")
+	flag.Usage = func() {
+		cmdName := path.Base(os.Args[0])
+		w := flag.CommandLine.Output()
+		fmt.Fprintf(w, "Usage:\n  %s [FLAGS]... IN_FILE... OUT_FILE\n\n", cmdName)
+		fmt.Fprintf(w, "Collects signals for each project repository listed.\n")
+		fmt.Fprintf(w, "IN_FILE must be either a file or - to read from stdin.\n")
+		fmt.Fprintf(w, "OUT_FILE must be either be a file or - to write to stdout.\n")
+		fmt.Fprintf(w, "\nFlags:\n")
+		flag.PrintDefaults()
+	}
+}
+
+func main() {
+	flag.Parse()
+
+	logger := log.New()
+	logger.SetLevel(logFlag.Level())
+
+	// roundtripper requires us to use the scorecard logger.
+	scLogger := sclog.NewLogrusLogger(logger)
+
+	if flag.NArg() < 2 {
+		logger.Error("Must have at least one input file and an output file specified.")
+		os.Exit(2)
+	}
+	lastArg := flag.NArg() - 1
+
+	// Open all the in-files for reading
+	readers := []io.Reader{}
+	consumingStdin := false
+	for _, inFilename := range flag.Args()[:lastArg] {
+		if inFilename == "-" && !consumingStdin {
+			logger.Info("Reading from stdin")
+			// Only add stdin once.
+			consumingStdin = true
+			readers = append(readers, os.Stdin)
+			continue
+		}
+		logger.WithFields(log.Fields{
+			"filename": inFilename,
+		}).Debug("Reading from file")
+		f, err := os.Open(inFilename)
+		if err != nil {
+			logger.WithFields(log.Fields{
+				"error":    err,
+				"filename": inFilename,
+			}).Error("Failed to open an input file")
+			os.Exit(2)
+		}
+		defer f.Close()
+		readers = append(readers, f)
+	}
+	r := io.MultiReader(readers...)
```

Comment:
nit: this function is very long. Does it make senes to break up this reader aggregation part into its own helper function and call it here instead?

Other parts of this function can probably be broken up too.

Reply:
Agreed.

For now though I'd prefer to leave it as-is. Moving the reader logic to a separate function complicates the file closing behaviour.